### DISCUSSION TITLE
Tapo C100 documentation ?

### DISCUSSION THREAD

**Discussion by silah** [themactep/thingino-firmware#400] [Q&A]
Hello

Sorry if I am missing something obvious. I have a Tapo C100 camera that I would like to try Thingino on. On the project website, the link to the firmware is dead. The Wiki doesn't mention it at all and I have googled, checked youtube, checked what I could find in this repo but there seems to be absolutely no guides / indications of how to flash Thingino on C100.

Am I missing obvious something somewhere ? Does anyone have a link to documentation or something ?

Thanks

**Comment by sstepansky** [comment]
Hi!
There are multiple ways to install a new firmware.
You can check the wiki page [here](https://github.com/themactep/thingino-firmware/wiki/Installation).

Maybe the link on the website is broken because a new tapo c100 variant has been added recently. Now there's the t31l and the t23n variants with different sensor and wifi chip.
Check the Github [releases page](https://github.com/themactep/thingino-firmware/releases).

**Reply by silah** [reply]
Yea I read the Wiki page but I guess you gotta really know your embedded devices before getting started with this.

I had kinda hoped for more of a step by step to be honest, something for people who haven't done much with embedded devices before but wanted to liberate their cameras. 

Only part I was comfortable doing was the SD card route and that didn't work, but no instructions on how to troubleshoot or where to look (and for what). The camera did attempt to update, but it didn't complete.

Im not complaining. I was just hoping that something a little more "beginner friendly" was out there but I get that its a lot of effort and I don't expect anyone to go out of their way to create that.

**Reply by sstepansky** [reply]
Yes, some knowledge of your device is needed. At least SoC, sensor and wifi chip. Vendors tend to change components without notice, so there may be more than one variant of the same model.

>The camera did attempt to update, but it didn't complete.

What have you tried? How do you know it did attemp to update?

**Reply by silah** [reply]
Yea I get that I will have to make sure the firmware is compatible with all the HW alright which is fine, I just wish there was a little more of a step-by-step.

I suppose I can't be sure if it was updating, but according to the manual a particular kind of flashing of the LED indicates a firmware update is taking place, and that kind of flashing happened when I went with naming the firmware autoupdate-full.bin  on an SD card

**Reply by sstepansky** [reply]
Well, I tested the _magic filename_ method outlined [here](https://web.archive.org/web/20240716193950if_/https://static.tp-link.com/upload/manual/2023/202312/20231204/How%20to%20manually%20upgrade%20the%20firmware%20of%20the%20Tapo%20Camera%20with%20a%20microSD%20card.pdf) and it fails due to a bug in the code... At least in my particular version...
I guess they don't bother fixing the bug because they want you to use the app...

So I guess in order to install some other firmware at least you'll need to open the camera housing and flash the rom chip or get a serial console to interact with the bootloader.

Do you have a serial adapter? And a programmer? [This one](https://github.com/themactep/thingino-firmware/wiki/CH341A-Programmer) can do both, is cheap and is well tested.

**Reply by silah** [reply]
Ill try that way on mine, but failing that I'll pick up a CH341A and give that a go. Thanks

**Comment by drew5494** [comment]
Hi I have a  TPLINK c100 with a T23.  I see it's supported on the website but no documentation in the wiki for the C100.  Any idea how I can go about  installing the firmware?

**Reply by drew5494** [reply]
![IMG_4386](https://github.com/user-attachments/assets/e6e967cf-3a84-4fff-9365-bf02b6444030)

**Reply by sstepansky** [reply]
Hi! The easiest way is to use a CH341A programmer. The method is analogous to that described in thr videos here https://github.com/themactep/thingino-firmware/wiki/Camera:-Jooan-A6M

**Comment by cawmit** [comment]
Waiting for a CH341A programmer to come in. Here is the bootlog from one of these, it's rather verbose. My camera is pretty much the same as @drew5494 except for the sticker 

<img width="1280" height="1279" alt="image1" src="https://github.com/user-attachments/assets/710dd2bc-d3d1-4f41-af77-ce01c0741e6a" />

[boot.log](https://github.com/user-attachments/files/22025660/boot.log)

This is the bootlog for a device that had already been setup with an account and connects to my home wifi 

Haven't tried interrupting the boot process yet etc

**Comment by cawmit** [comment]
Been digging around in a ROM dump for a bit now, inspected both the SD card upgrade route and openings in the exposed web services during the onboarding state. At this point I'm quite confident in saying that there is no way to update the firmware through the sd card port and the only way to upgrade the firmware at a minimum requires opening it up. (happy to be wrong 😄)  

Two ways to update;  
- desolder a few pins on the ROM chip and use a CH341A programmer   
- use the UART port and a SD card, which I recommend    

Just clipping on to the ROM chip will power the SoC and cause the CH341A programmer to not work

Via uart you can stop uboot by typing `slp` during the "Autobooting in 1 seconds" message, or let it boot and log in to the linux shell with `root` and `slpingenic`. The built-in tools want a specific RSA-signed image and a fixed partition layout. The RSA signing check can be disabled but Thingino doesn’t match that the partition layout, so I flashed from U-Boot instead.

These instructions have _only_ been tested V5 C100 T23N hardware and the current Thingino C100 T23N firmware download. These actions could easily brick your device do not use them if you do not have a full ROM backup and/or do not know what you are doing; 

0. create a fat32 sd card and put the c100 t23n .bin file on it
1. interrupt u-boot by quickly during typing `slp` when "Autobooting in 1 seconds" message shows up  
2. make sure the sd card is inserted, run `mmc rescan`  
3. run `sf probe 0` 
4. validation step, run `fatls mmc 0:1`, in the response you should see the .bin file for the firmware upgrade on the sd card   
6. load the new firmware into memory, run `fatload mmc 0:1 0x82000000 upg.bin` (replace upg.bin with your filename)  
7. **danger** step, we're going to erase the existing rom, everything before this was reversible, this is not, if anything after this step fails, the camera is bricked and needs a ROM programmer to be recovered, run `sf erase 0 0x800000`
8. run `sf write 0x82000000 0 ${filesize}` this writes the .bin file we loaded into memory on to the ROM chip we just erased
9. after this you are done and optionally can run these validation steps to make sure the firmware upgrade was successful ;   
9.1. run `sf read 0x83000000 0 ${filesize}` it should should return "SF: 8388608 bytes @ 0x0 Read: OK"  
9.2. run `cmp.b 0x83000000 0x82000000 ${filesize}` it should return "Total of 8388608 byte(s) were the same"   
10. run `reset` to reboot the camera, it will now load into Thingino. From here you can connect to the cameras wifi network and follow the [documentation to configure the camera](https://github.com/themactep/thingino-firmware/wiki/Configuration:-Wi%E2%80%90Fi-Access#captive-portal)

**Comment by paintballkev812** [comment]
I purchased a 4 pack of the Tapo C110 v3.6 (which appear to be identical to the C100 5.0 hardware wise) to load the thingino firmware on them. Unfortunately, I have been unable to get a uart connection. I've attempted the connection using an FTDI232 as well as a CH341a programmer, but unable to get any output.

I fear the issue is that the firmware that was shipped on these have disabled uart, but wanted to post here to see if I am doing something wrong or any suggestions on what to do next.

I've been connecting to the pads following camwit's photo using a [pogo pin adapter](https://www.printables.com/model/122176-flash-adapter-with-pogo-test-pins) to connect to TX,RX, and GND. The pin appear to be making solid contact and tested with a multimeter to confirm continuity. I am using putty for the terminal with a baud rate of 115200 and have turned off the flow control. with the usb adapter, I will connect the barrel plug to power the camera. The terminal stays open, but I am not getting output to the terminal. I have also swapped the TX and RX pins around to make sure I didn't mix them up, but still no result. I've also tried using a multimeter to measure voltage between gnd and TX/RX during the camera startup, but did not see any voltage readings.

On another github topic (on OpenIPC, I believe), I saw reference that some resistors near the pads were missing (likely in an attempt to prevent uart access), but my camera does have the resistors in place (see images below).

Would appreciate any insight.

![C110 Back](https://github.com/user-attachments/assets/4ac75794-5f2e-41d4-9bba-694183abbbf8)
![C110 Front](https://github.com/user-attachments/assets/7b2612fa-a849-45f8-b1ae-498a33de216f)

**Reply by sstepansky** [reply]
Yes, you are missing the resistors. Put a blob of solder to close the connection.

C100 and C110 are almost identical except for the sensor. Yours should be identical to C110 v3, not C100. Once you get a serial connection, check the bootlog to confirm the sensor.

**Reply by paintballkev812** [reply]
Thanks for this! I had felt the solder bumps and assumed that was the resistor. After reading your comment, I took a close up photo and can confirm there are two tiny resistors missing. I will try to put a blob of solder and try again.

![Missing Resistors](https://github.com/user-attachments/assets/099ff38d-6d12-4b47-a472-fdf328541095)
